jetty-websocket/jetty-websocket-tests/src/test/java/org/eclipse/jetty/websocket/tests/client/ConnectFutureAbortTest.java

```diff
@@ -0,0 +1,284 @@
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.hamcrest.Matchers.is;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class ConnectFutureAbortTest
+{
+    private Server server;
+    private WebSocketClient client;
+
+    @FunctionalInterface
+    public interface Configuration
+    {
+        void configure(NativeWebSocketConfiguration configuration);
+    }
+
+    public void start(Configuration configuration) throws Exception
+    {
+        server = new Server();
+        ServerConnector connector = new ServerConnector(server);
+        server.addConnector(connector);
+
+        ServletContextHandler contextHandler = new ServletContextHandler();
+        contextHandler.setContextPath("/");
+        server.setHandler(contextHandler);
+
+        NativeWebSocketServletContainerInitializer.configure(contextHandler, (context, container) ->
+            configuration.configure(container));
+        contextHandler.addFilter(WebSocketUpgradeFilter.class, "/", EnumSet.of(DispatcherType.REQUEST));
+        server.start();
+
+        client = new WebSocketClient();
+        client.start();
+    }
+
+    @AfterEach
+    public void stop() throws Exception
+    {
+        if (client != null)
+            client.stop();
+        if (server != null)
+            server.stop();
+    }
+
+    @Test
+    public void testAbortDuringCreator() throws Exception
+    {
+        final CountDownLatch enteredCreator = new CountDownLatch(1);
+        final CountDownLatch exitCreator = new CountDownLatch(1);
+        start(c ->
+        {
+            c.addMapping("/", (req, res) ->
+            {
+                try
+                {
+                    enteredCreator.countDown();
+                    exitCreator.await();
+                    return new EventSocket.EchoSocket();
+                }
+                catch (InterruptedException e)
+                {
+                    throw new IllegalStateException(e);
+                }
+            });
+        });
+
+        CloseTrackingEndpoint clientSocket = new CloseTrackingEndpoint();
+        Future<Session> connect = client.connect(clientSocket, WSURI.toWebsocket(server.getURI()));
+
+        // Cancel the future once we have entered the servers WebSocketCreator (after upgrade request is sent).
+        assertTrue(enteredCreator.await(5, TimeUnit.SECONDS));
+        assertTrue(connect.cancel(true));
+        assertThrows(CancellationException.class, () -> connect.get(5, TimeUnit.SECONDS));
+        exitCreator.countDown();
+        assertFalse(clientSocket.openLatch.await(2, TimeUnit.SECONDS));
+
+        // Strange that onError() is called but onOpen() is never called. (This would not be the case in Jetty-10).
```

Comment:
Strange comment :)
Why would `onOpen()` even be called, if the `connect()` has been canceled?

Reply:
I didn't mean that `onOpen()` should be called, but that `onError()` should only be called if `onOpen()` has been called previously. I just wasn't expecting both of them to be notified of the failure, only the future. I will remove the comment.